Pomodoro timer

fyyhr opened the pull request:
I think this should work. Working on some test cases, but I figured you guys can look at this to setup the UI.

Let me know if you need more info from the class somewhere.

fyyhr committed:
fix'd .classpath

fyyhr committed:
created pomodoro class

fyyhr committed:
Created PomdoroTimer class and part of executor clas

TODO: Finish PomodoroExecutor and add unit tests

fyyhr committed:
Refactoring, added more TODOs

TODO: secure pomodoro timer class methods
TODO: Finish pomodoro execution countdown

+ generic delay function, delay_sec, that waits about 1 second, handled in new thread
= baseline of countdown function, decrementing counter that can be read independently

fyyhr committed:
Completed pomodoro countdown timer

+ finished pomodoro countdown
= refactored package to start with lowercase

mckuok commented on src/Pomodoro/exec/PomodoroExecutor.java:
```diff
@@ -0,0 +1,233 @@
+	private int currentPomodoro;					//what number pomodoro is this, starts at 1
+	
+	/* booleans indicate what timer is currently going and if all pomodoros have finished */
+	private boolean Pomodoro;
```
For states, it should be presented using enums, especially states are exclusive

mckuok replied:
Using booleans it is very easy to forget to set it back to true / false

mckuok commented on src/Pomodoro/exec/PomodoroExecutor.java:
```diff
@@ -0,0 +1,233 @@
+		temp += this.getTimer().getBreakSec();
+		this.setBreakRemaining(temp);
+		/* set long break timer */
+		temp = this.getTimer().getLongBreakMin();
```
DRY, duplicated logics like this should be handled by a method

mckuok commented on src/Pomodoro/PomodoroTimer.java:
```diff
@@ -0,0 +1,174 @@
+	 * @param longBreakFreq how many breaks between long breaks (ex. value of 3 means 2 normal 
+	 *     breaks and the 3rd is a long break)
+	 */
+	public PomodoroTimer(int mins, int sec, int pomodoros, int breakMin, 
```
Since the time interval is most likely fixed, we don't want to ask the user to new input everytime. Even if we do, at least we want to give them a default that is mapped from the previous setup. So you would want to integrate some of theses settings to ProjectConfig for persistence.

fyyhr replied:
So should that persistence be handled in the pomodorotimer class or the proj config? I thought it might be easier to handle that in the project config, that way this class wouldn't need to access external memory

mckuok replied:
The variables needed to be stored on disk (number of pomodoro, frequencies etc.) will need to be in ProjectConfig, and you can initialize them with  defaults when you are declaring the variables (You can look at how GitConfig is done in ProjectConfig and use it as an example). 

You should do it in ProjectConfig because ConfigManager should be smarter enough to pick up the changes you made in ProjectConfig. You might also want to write a unit test on the Pomodoro part of persistence (you can look at the existing one as example). to make sure it works.

TBH right now ProjectConfig is too crowded, I will most likely extract GitConfig and TrelloConfig to its own class, but we can do that later. 

The constructor of this PomodoroTimer can stay as is. What we will end up doing is most likely read the config stored on disk using ConfigManager, and pass those settings to your PomodoroTimer

mckuok commented on src/Pomodoro/exec/PomodoroExecutor.java:
```diff
@@ -0,0 +1,233 @@
+	 */
+	
+	public void countdown() throws InterruptedException {
+		while (this.getCurrentPomodoro() != this.getTimer().getPomodoros()) {
```
I would want a way to be able to pause the timer while keeping its current state to offer a greater control to the user

mckuok commented on src/Pomodoro/PomodoroTimer.java:
```diff
@@ -0,0 +1,174 @@
+		 * Check if the seconds are proper and increment minutes if they are not. If anything is 
+		 * negative, set to 0. If pomodoros or long_break_freq is negative, set to 1
+		 */
+		while (sec >= 60) {
```
Since in the executor you convert everything back to seconds, why don't you just convert the minute to seconds here to save one layer of conversion? 

You can still take the min in the constructor, but internally you shouldn't need the min at all

mckuok commented on src/Pomodoro/exec/PomodoroExecutor.java:
```diff
@@ -0,0 +1,233 @@
+		this.timer = timer;
+	}
+	
+	public void setTimeRemaining(long time) {
```
Arbitrary method name, not sure what is Time referring to

mckuok commented on src/Pomodoro/exec/PomodoroExecutor.java:
```diff
@@ -0,0 +1,233 @@
+		} else if (this.isLongBreak()) {
+			this.setCurrentTime(this.getLongBreakTime());
+		} else if (this.isFinished()) {
+			/* Hey, you finished */
```
To avoid things like this, you might want to make this class abstract, and have an abstract method like:
public abstract void isFinished();

fyyhr committed:
Removed redundancies and shortened code

+ Added a pause boolean to handle timer pausing. Incomplete.
- Converted booleans to enum
- Removed redundant calling of temp *= 60 to convert minutes to seconds
= Moved minutes to second conversions to the class constructor from the executor initializer
= Created a void function isFinished() to handle what happens when the last pomodoro finishes. Incomplete

TODO: Implement Interrupts to handle pausing/unpausing
TODO: Complete the isFinished function

fyyhr committed:
Implemented pause to interrupt in the delay_sec() function

+ Added catch interruptedexception in delay_sec() that will set the pause booelean to true, pausing the timer

fyyhr committed:
Refactor time holding variable names and their functions to remove ambiguity.

fyyhr committed:
Bugfix - Incorrect variable names being called

fyyhr committed:
removed false TODO statement

fyyhr committed:
Merge branch 'master' of https://github.com/mckuok/TrelloGitHubIntegration into PomodoroTimer
